src/main/kotlin/com/londogard/textgen/languagemodels/LanguageModel.kt

```diff
@@ -0,0 +1,101 @@
+    val n: Int
+) {
+    fun getReverseDictionary(): InternalReverseVocabulary = dictionary.entries.associateBy({ it.value }) { it.key }
+
+    fun getLanguageModel(): InternalLanguageModel = internalLanguageModel
+
+    @ExperimentalSerializationApi
+    fun dump(absolutePath: String): Unit = SerializerUtil.encodeLanguageModel(absolutePath, this)
+
+    companion object {
+        @ExperimentalSerializationApi
+        fun dump(languageModel: LanguageModel, absolutePath: String): Unit =
+            SerializerUtil.encodeLanguageModel(absolutePath, languageModel)
+
+        @ExperimentalSerializationApi
+        fun load(absolutePath: String): LanguageModel = SerializerUtil.decodeLanguageModelByAbsPath(absolutePath)
+
+        fun fit(
+            documents: List<String>,
+            n: Int,
+            tokenizer: Tokenizer = SimpleExtensibleTokenizer(whitespaceRegex = " "),
+            keepMinFreq: Int = 0, // 0 = at least one occurrence, 1 = at least two occurrences, etc... Applied to all ngrams
+            preprocessing: List<(String) -> String> = emptyList(),   // Example: listOf(TextNormalizer::normalize, String::toLowerCase)
+        ): LanguageModel {
+            val preprocessedDocs = documents
+                .map { subDoc -> preprocessing.fold(subDoc) { acc, preprocessor -> preprocessor(acc) } }
+            val tokens = preprocessedDocs.map(tokenizer::split)
+            val reverseDict = tokens.flatten().toHashSet().let { uniqueTokens ->
+                HashMap<String, Int>(uniqueTokens.size).apply { putAll(uniqueTokens.zip(uniqueTokens.indices)) }
+            }
+            val tokensList = tokens.map { t -> t.mapNotNull(reverseDict::get) }
+
+            var previousMatch = true
+            val ngramMap = (1..n)
+                .fold(emptyMap<List<Int>, Int>()) { acc, ngram ->
+                    when {
+                        ngram <= 2 || previousMatch ->
+                            acc.merge(tokensList
+                                .filter { docToken -> docToken.size >= ngram }
+                                .flatMap { docToken ->
+                                    docToken
+                                        .indices
+                                        .mapNotNull { index -> // Even though autobox applies to List<E> we save space here because of views.
+                                            val topIdx = (index + ngram).coerceAtMost(docToken.size)
+
+                                            if (ngram <= 2 || (acc.contains(docToken.subList(index, topIdx - 1))))
+                                                docToken.subList(index, topIdx)
+                                            else null
+                                        }
+                                }
+                                .groupingBy { it }
+                                .eachCount()
+                                .filter { (key, value) -> key.size < ngram || value > keepMinFreq }
```

Comment:
If we want key.size < ngram depends on question above, do we want 3-gram if sentence is ["hello", "world"] as it will match "hello","world", but already has on bigram?